### ISSUE TITLE
Possible to avoid theano.sandbox import during lasagne / layers import?

### ISSUE THREAD

**Issue by astooke** [Lasagne/Lasagne#775]
Hi,
I am using Lasagne in a multi-GPU machine, in which the master process initializes with `theano.config.device=cpu`, then sub-processes are started, and after that each one calls e.g. `theano.sandbox.cuda.use('gpu0')` (or whatever number gpu it is assigned) and then builds its own Theano variables and functions.  The problem I'm having is that this fails if Lasange has been imported in the master process before forking.  I think the reason is the few places within `layers` where `theano.sandbox.cuda` is imported during Lasagne initialization.  

I've gotten around the problem in my case by moving all Lasagne imports inside the initialization of the objects that use them, and I'm careful not to instantiate these until after forking.  I would prefer to be able to leave all the Lasagne imports in the file headers, and have everything work on the GPU as long as I'm careful not to build any Theano variables until after calling `theano.sandbox.cuda.use()`, which is how it can be done with imports of `theano` and `theano.tensor`.

I think this could be achieved by applying the workaround I just described to the few Lasagne layers with the culprit import statements in them.  Does this sound reasonable?  Of course I could do this to my local version (haven't actually tried it yet) but that's no good for sharing code.

Thanks,
Adam

p.s. I've also found it to break when importing `theano.sandbox.rng_mrg`, so it might just be about getting anything from the sandbox.  I'm on bleeding edge Theano and Lasagne and Ubuntu 16.04.

**Comment by f0k** [comment]
> The problem I'm having is that this fails if [...]

Can you post the error message here, and possibly a short demonstration script that doesn't do much besides exposing the error? Maybe we can find an easier way than hiding all the `sandbox` imports (a way that involves modifying Theano).

**Comment by astooke** [comment]
The error when importing `lasagne` or `theano.sandbox.cuda` is:

```
>>> ERROR (theano.sandbox.cuda): ERROR: Not using GPU. Initialisation of device 1 failed:
>>> Unable to get properties of gpu 1: initialization error
```

I tried chasing this error a little bit, but was not successful at troubleshooting.  

And here is a little demo script to recreate it (note you'll need two GPUs).  Turns out importing `theano.sandbox` is OK, but not certain things from it:

```
import theano
import theano.tensor as T
import theano.sandbox
import multiprocessing as mp
import time

# ANY OF THESE CAUSES THE ERROR
# import theano.sandbox.cuda
# from theano.sandbox.rng_mrg import MRG_RandomStreams
# import lasagne

x = T.scalar('x', dtype='float32')  # just exercising theano.tensor

def target():
    import theano.sandbox.cuda
    print("target about to use")
    theano.sandbox.cuda.use('gpu1')
    print("target is using")
    import lasagne
    time.sleep(15)  # (make it long enough to see master hold GPU simultaneously)
    print("target is exiting")

p = mp.Process(target=target)
p.start()
import theano.sandbox.cuda
print("master about to use")
theano.sandbox.cuda.use('gpu0')
print("master is using")
import lasagne
p.join()
print("master is exiting")
```

When it's working properly, the full output is something like this:

```
>>> master about to use
>>> target about to use
>>> Using gpu device 0: GeForce GTX 1080 (CNMeM is enabled with initial size: 50.0% of memory, cuDNN 5105)
>>> master is using
>>> Using gpu device 1: GeForce GTX 1080 (CNMeM is enabled with initial size: 50.0% of memory, cuDNN 5105)
>>> target is using
>>> target is exiting
>>> master is exiting
```

**Comment by f0k** [comment]
Interesting. I don't think this is intended. `theano.sandbox.rng_mrg` imports `theano.sandbox.cuda` as well: https://github.com/Theano/theano/blob/f22f449/theano/sandbox/rng_mrg.py#L25. And `import lasagne` triggers `import theano.sandbox.rng_mrg`.
So there are at least two solutions: Avoid `theano.sandbox.rng_mrg` importing `theano.sandbox.cuda`, or avoid the `theano.sandbox.cuda` import messing up anything for a subsequently forked process. The latter would be preferable.
We should try to get closer to the underlying cause and report this upstream as a Theano issue.

**Comment by astooke** [comment]
Good idea!  I hadn't considered that this might not be the intended behavior of importing `theano.sandbox.cuda`.  I won't have time in the near future to dig deeper, but I'm happy to post as a Theano issue...I'll do that in a day or two if you don't find out anything new before then.

**Comment by f0k** [comment]
Okay, the problem is that `import theano.sandbox.cuda` calls `gpu_init()` without parameters just to check whether CUDA is available at all. It does not set an active device yet, but it includes a call to `cudaGetDeviceCount()`: https://github.com/Theano/Theano/blob/c89d973/theano/sandbox/cuda/cuda_ndarray.cu#L3194
This implicitly creates a CUDA context used for all subsequent CUDA calls. Now when you fork the process afterwards, the child will inherit the CUDA context and try to use it when you call `use("gpu1")`, failing at `cudaGetDeviceProperties` in `gpu_init()`: https://github.com/Theano/Theano/blob/c89d973/theano/sandbox/cuda/cuda_ndarray.cu#L3217
The reason is that a CUDA context cannot be forked.

A solution might be to record the ID of the process that created the context (i.e., before setting `g_gpu_context_active` from 0 to 1), and create and activate a new context at the beginning of `gpu_init()` if we find there already is a context for a different process ID. The functions for managing contexts are in the Driver API: http://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html
If you get to a working solution, send a PR to Theano. (This is for the old backend, which is slowly getting phased out, but I think they will still accept it. I don't know if the new backend has the same problem.) Also please report back how things work out!

**Comment by astooke** [comment]
Good find!  I had gotten as far as `cudaGetDeviceProperties` in `gpu_init()` but hadn't sorted out what was breaking that.  I'm out for all of December, so I likely won't be able to do this until January.  Sounds like you already have a solution in mind, so if you want to do it and submit before then, that'd be fine.  Otherwise I'll definitely do it in January because this will prevent having to modify other existing code for my team.

Also a good idea to look into the new backend...I haven't tried that or read anything about it yet.

**Comment by f0k** [comment]
> Sounds like you already have a solution in mind, so if you want to do it and submit before then, that'd be fine. Otherwise I'll definitely do it in January because this will prevent having to modify other existing code for my team.

Would be good if you could tackle this yourself! It's a very specific problem and I don't have the time either... By the way, getting the PID in a platform-independent way may be a nuisance, you may also try to import [`pycuda.driver`](https://documen.tician.de/pycuda/driver.html) in your forked process and create a new context there. I think theano.sandbox.cuda does not explicitly create a context anywhere, so it will just use whatever is the current context.

> Also a good idea to look into the new backend...I haven't tried that or read anything about it yet.

Basically you [install libgpuarray and pygpu](http://deeplearning.net/software/libgpuarray/installation.html), and then use `device=cuda` instead of `device=gpu`. Things in Lasagne that explicitly use `theano.sandbox.cuda` won't work then, but those are usually not needed anyway.

**Comment by astooke** [comment]
Just started using the new Theano GPU backend, and it supports use of multiple GPU's from within one python process, which solves my problem.  

To have multiple subprocesses using GPU(s), now Theano itself must not be imported until after the subprocesses are forked, as per your point about forking CUDA contexts.  (This is using either `THEANO_FLAGS="contexts=dev0->cuda0;..etc"` or `"device=cuda"`.)   I haven't seen yet a way to delay initialization of GPUs until sometime after importing Theano with the new backend...less important with multi-GPU support but could still be handy.  Interestingly, multiple different python processes can use the same GPU now, too. 

Clearly the new backend is the way forward, and this issue is not about Lasagne, so, closing this one.  Thanks for the help.

**Comment by astooke** [comment]
Actually, this is still a problem, since we're using multiprocessing for multiple GPUs after-all to ensure concurrency for data parallelism.  Theano issue [raised here](https://github.com/Theano/Theano/issues/5557).
